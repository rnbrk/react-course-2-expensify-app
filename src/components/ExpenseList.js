import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
    )}
    {}
  </div>
);

// MAP REDUX STATE TO THE PROPS (CALLBACK)
// WILL UPDATE EVERY TIME SOMETHING CHANGES TO THE STATE
const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

// CONNECT TO THE STORE WITH
// FUNCTION (returns)-> HIGHER ORDER COMPONENT (returns)-> REACT COMPONENT
export default connect(mapStateToProps)(ExpenseList);
