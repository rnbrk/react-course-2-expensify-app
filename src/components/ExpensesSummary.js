import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getExpensesTotal from '../selectors/getExpensesTotal';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => (
  <div>
    <p>{`Viewing ${expenseCount} expenses`}</p>
    <p>{`Totaling ${numeral(expensesTotal).format('$0,0.00')}`}</p>
  </div>
);

const mapStateToProps = state => ({
  expenseCount: state.expenses.length,
  expensesTotal: getExpensesTotal(state.expenses) / 100
});

export default connect(mapStateToProps)(ExpensesSummary);
