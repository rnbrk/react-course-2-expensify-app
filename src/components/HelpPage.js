import React from 'react';

const HelpPage = () => (
  <div>The Help component!</div>
);

export default HelpPage;
