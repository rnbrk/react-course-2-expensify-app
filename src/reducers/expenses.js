const expensesReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        // Copies expenses array and adds an extra expense
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map(expense =>
        expense.id === action.id
          ? {
              ...expense,
              ...action.updates
            }
          : expense
      );
    case 'SET_EXPENSES':
      return [...action.expenses];
    default:
      return state;
  }
};

export default expensesReducer;
