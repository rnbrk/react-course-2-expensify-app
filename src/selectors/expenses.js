import moment from 'moment';

// Filters and sorts the expenses according to the supplied filters
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) =>
  expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return b.createdAt - a.createdAt;
      }
      if (sortBy === 'amount') {
        return b.amount - a.amount;
      }
    });

export default getVisibleExpenses;
