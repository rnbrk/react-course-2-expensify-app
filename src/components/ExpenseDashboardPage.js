import React from 'react';
import ExpenseListConnect from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpensesSummary />
    <ExpenseListConnect />
  </div>
);

export default ExpenseDashboardPage;
