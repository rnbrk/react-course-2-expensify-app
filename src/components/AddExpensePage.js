import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    // dispatch(addExpense(expense));
    this.props.startAddExpense(expense); // Replaces line above
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

// This makes sure we can test this component easier
const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
