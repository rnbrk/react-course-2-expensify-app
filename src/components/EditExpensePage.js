import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button type="button" onClick={this.onClick}>
          Remove
        </button>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
