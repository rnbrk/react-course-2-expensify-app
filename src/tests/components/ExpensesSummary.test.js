import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import getExpensesTotal from '../../selectors/getExpensesTotal';

test('Should display correctly without expenses', () => {
  const wrapper = shallow(<ExpensesSummary />);
  expect(wrapper).toMatchSnapshot();
});

test('Should display correctly with expenses', () => {
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={expenses.length}
      expensesTotal={getExpensesTotal(expenses) / 100}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
