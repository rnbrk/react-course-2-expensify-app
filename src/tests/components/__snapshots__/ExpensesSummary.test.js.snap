// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should display correctly with expenses 1`] = `
<div>
  <p>
    Viewing 3 expenses
  </p>
  <p>
    Totaling $768.89
  </p>
</div>
`;

exports[`Should display correctly without expenses 1`] = `
<div>
  <p>
    Viewing undefined expenses
  </p>
  <p>
    Totaling $0.00
  </p>
</div>
`;
